validator-client-oapi.yml

```
@@ -0,0 +1,247 @@
+    and "Accept: application/json" headers.
+
+    All sensitive routes are to be authenticated with a token. This token should be provided by the user via a secure channel:
+      - Log the token to stdout when running the binary with the validator client enabled
+      - Read the token from a file available to the binary
+  version: "Dev - Eth2Spec v1.0.1"
+  contact:
+    name: Ethereum Github
+    url: https://github.com/ethereum/keymanager-APIs/issues
+  license:
+    name: "Apache 2.0"
+    url: "https://www.apache.org/licenses/LICENSE-2.0.html"
+
+servers:
+  - url: "{server_url}"
+    variables:
+      server_url:
+        description: "key manager API url"
+        default: "https://public-mainnet-node.ethereum.org"
+
+tags:
+  - name: RemoteKeymanager
+    description: Set of endpoints for key management.
+
+paths:
+  /eth/v1/remotekey:
+    get:
+      operationId: ListRemoteKeys
+      summary: List Remote Keys.
+      description: |
+        List all remote validating pubkeys known to this validator client binary
+      security:
+        - bearerAuth: []
+      tags:
+        - RemoteKeymanager
+      responses:
+        "200":
+          description: Success response
+          content:
+            application/json:
+              schema:
+                title: ListRemoteKeysResponse
+                type: object
+                required: [data]
+                properties:
+                  data:
+                    type: array
+                    items:
+                      $ref: "#/components/schemas/SignerDefinition"
+        "401":
+          $ref: "#/components/responses/Unauthorized"
+        "403":
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalError"
+
+    post:
+      operationId: ImportRemoteKeys
+      summary: Import Remote Keys.
+      description: |
+        Import remote keys for the validator client to request duties for.
```

how is this intended to be used? does this just add to the static list that you started the validator with? does it error if you started with a url endpoint? ( you can't have both a url and a static list)

---

I think this can be implementation dependent. If an implementation can't run with static + dynamic keys then it should error if it's in static mode. If it can accommodate both or its in dynamic mode/dynamic only, then it won't error.

---

>  If it can accommodate both or its in dynamic mode/dynamic only, then it won't error.
so the validator would then call the remote signer to add /remove the key using the new standard apis that @joaquim-verges implemented on the web3signer? I guess this to me calls into question how does the validator authenticate to the remote signer to do this, and if that's a security risk for validators to be able to call the remote signer itself to add and remove dynamic keys... it feels like the intended usage of the web3signer is to not allow this from the validator side.

---

> so the validator would then call the remote signer to add /remove the key using the new standard apis that @joaquim-verges implemented on the web3signer?

No. The assumption is that those keys need to have been added to web3signer or the remote signer separately. That's why the validator client doesn't need access to the keystore when adding a remote signing validator.

---

The use-case looks something like this:

_As a DAppNode user I want to add a new keystore to my validator client_

_Steps:_

1. _DAppNode imports the keystore (with password) to web3signer using the keymanager API_
2. _DAppNode adds the public key as a remote signed key to the validator client of choice (e.g. Prysm, Lodestar, etc)_

It also enables this use-case:

_As a DAppNode user I want to switch validator clients quickly_

_Steps:_

1. _DAppNode removes the remote signing public key from the previous validator client (e.g. Lighthouse)_
2. _DAppNode adds the same public key as a remote signer key to the new validator client (e.g. Teku)_

^Note how in the second use-case we don't need the keystore material again because it's already held by the remote signer. We're just reconfiguring validator clients.

---

This usecase definitely makes sense to me, i was thinking of the situation where you do something like you starting the validator with --validators-external-signer-public-keys= localhost:9000/api/v1/eth2/publicKeys where the validator is starting with a url defined dynamic list of public keys provided by the remote signer itself... and if the addition and removal should be added there. makes sense for adding and removing the way your usecase explains it for the list that lives with the validator itself.

---

Oh yeah, in that case all these APIs should probably return an error saying "remote signer definitions are automatically managed". AFAIK only Teku is in that boat currently. For Lighthouse I think we'd prefer to not implement the auto-reload functionality and have the user manage the set of active keys explicitly (with this API).

---

Teku doesn't automatically load keys - you have to manually specify them.  But one way you can manually specify them is by giving Teku a URL to load them from.  And that URL could be the Web3Signer endpoint that lists all the public keys it has.  Even then you have to send a `HUP` signal to explicitly tell us to reload the URL and pick up any new validators.

---

> Teku doesn't automatically load keys - you have to manually specify them. But one way you can manually specify them is by giving Teku a URL to load them from. 

oh wait let me try to understand this... does that mean the url only loads once then caches it? it's not like teku's internal proxies to the url each time? so if it reloads on the web3signer then you get an updated list?

> Even then you have to send a HUP signal to explicitly tell us to reload the URL and pick up any new validators.
so on the web 3 signer part you call the reload api but does this mean you need to call the validator too? ( related to previous comment)

---

In that mode of operation though you don't know when the user is going to `SIGHUP` and refresh the list of keys. I think you'd have to error on remove because a removed key could be brought back by a refresh. Adding keys seems like it would be fine.

---

> oh wait let me try to understand this... does that mean the url only loads once then caches it? it's not like teku's internal proxies to the url each time? so if it reloads on the web3signer then you get an updated list?

Correct, at startup Teku retrieves the URL and treats its content as if those public keys were specified on the CLI.  The only time it will download that URL again is if you send a `HUP` signal which is how you get Teku to load new validators (including looking for any new keystores in directories you specified with `--validator-keys`).

Adding and removing keys from Web3Signer has no effect on Teku.

There is no requirement at all that the URL you get Teku to load public keys from needs to be from Web3Signer. It's just that Teku tends to support loading content from URLs for a lot of things (eg `--initial-state` can take a file path to the state or a URL).

---

> In that mode of operation though you don't know when the user is going to `SIGHUP` and refresh the list of keys. I think you'd have to error on remove because a removed key could be brought back by a refresh. Adding keys seems like it would be fine though.

I think we'll need to think through the details. Sending a `HUP` to load new keys seems pretty equivalent to sending an API request to re-add the key so I'm not sure it's an issue.

Generally I don't recommend setting up Teku to automatically load every key Web3Signer has - it's much better to be more explicit about where the keys come from. But you might be pointing the URL to a custom key management system that controls which keys run where in your setup rather than web3signer - it's still just a URL to Teku.